core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,747 @@
+        # prediction so that it is classified as correct class.
+        self.occurrence = None
+
+        # Value of K in K nearest neighbours.
+        self.top = None
+
+        # sklearn.feature_extraction.text.CountVectorizer object. It converts
+        # text into a feature a vector.
+        self.count_vector = None
+
+    def to_dict(self):
+        """Returns a dict representing this classifier.
+
+        Returns:
+            dict. A dictionary representation of classifier referred as
+            'classifier_data'. This data is used for prediction.
+        """
+        fingerprint_data = {
+            index: {
+                'fingerprint': self.training_data[index]['fingerprint'],
+                'class': self.training_data[index]['class']
+            } for index in self.training_data
+        }
+
+        classifier_data = {
+            'KNN': {
+                'token_to_id': self.token_to_id,
+                'T': self.T,
+                'K': self.K,
+                'top': self.top,
+                'occurrence': self.occurrence,
+                'fingerprint_data': fingerprint_data
+            },
+            'SVM': classifier_utils.extract_svm_parameters(self.clf),
+            'cv_vocabulary': self.count_vector.__dict__['vocabulary_']
+        }
+        return classifier_data
+
+    # pylint: disable=too-many-locals
+    def train(self, training_data):
+        """Trains classifier using given training_data.
+
+        Args:
+            training_data: list(dict). The training data that is used for
+                training the classifier. The list contains dicts where each dict
+                represents a single training data group, for example:
+                training_data = [
+                    {
+                        'answer_group_index': 1,
+                        'answers': ['a1', 'a2']
+                    },
+                    {
+                        'answer_group_index': 2,
+                        'answers': ['a2', 'a3']
+                    }
+                ]
+        """
+        data = collections.OrderedDict()
+        count = 0
+        for answer_group in training_data:
+            for answer in answer_group['answers']:
+                data[count] = {
+                    'source': answer,
+                    'class': answer_group['answer_group_index']
+                }
+                count += 1
+
+        data, token_to_id = tokenize_data(data)
+
+        # No. of nearest neighbours to consider for classification using KNN.
+        top = int(math.ceil(math.sqrt(float(len(training_data)))))
+
+        # GRID search for best value of T and K. T and K are model parameters
+        # of winnowing algorithm.
+        best_t = T_MIN
+        best_k = K_MIN
+        previous_best_score = 0.0
+        for T in range(T_MIN, T_MAX):
+            for K in range(K_MIN, T + 1):
+                data = add_k_gram_hashes(data, token_to_id, K)
```

Comment:
Hmm dicts are mutable, so you need not return it from within the function. You can just modify it inside the function.